### PR TITLE
ERA-862 Rework reset

### PR DESCRIPTION
**Pull request by gaborh-da**

### CODE REVIEW

File: src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java [hunk 1]
```diff
@@ -76,19 +83,32 @@ public Builder damlRoot(Path damlRoot) {
       return this;
     }
 
-    public DamlScriptRunner build() {
+    public Builder parties(String[] parties) {
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:86]: maybe this?
```suggestion
    public Builder uniqueParties(String[] uniqueParties) {
```

File: src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java [hunk 2]
```diff
@@ -76,19 +83,32 @@ public Builder damlRoot(Path damlRoot) {
...
+      this.parties = parties;
+      return this;
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: Some assert here would be very developer friendly to ensure that these party names weren't used before.
We can implement in a separate PR, let's not block this. Made a note on the jira.

File: src/test/java/com/digitalasset/testing/PingPongIT.java [hunk 3]
```diff
@@ -72,76 +74,83 @@ public void testCreate() throws InvalidProtocolBufferException {
...
+        record(field("sender", alice), field("receiver", bob), field("count", int64(2)));
 
-    ledger().getCreatedContractId(BOB, pingTemplateId(), recordMatcher, ContractId::new);
+    ledger().getCreatedContractId(bob, pingTemplateId(), recordMatcher, ContractId::new);
   }
 
   @Test
   public void testObservationWithoutMatcher() throws InvalidProtocolBufferException {
-    ledger().getCreatedContractId(BOB, pingTemplateId(), ContractId::new);
+    Party bob = getUniqueParty("Bob");
```
Comment (richardkapolnai-da) [inline src/test/java/com/digitalasset/testing/PingPongIT.java:84]: why is this necessary? can you remove?

Comment (gaborh-da) [reply src/test/java/com/digitalasset/testing/PingPongIT.java:84]: Yes, this should not be here, I should have removed all these

File: src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java (hunk 2 above)
Comment (gaborh-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: One can only check whether parties contains a duplicate or not - is that enough?

Comment (richardkapolnai-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: I think we need more. We need a mechanism to raise error when the developer forgot to call the obfuscator and simply passes "Bob" in multiple tests:
```
testA()
...
parties(["Bob"]);

testA()
...
parties(["Bob"]); // raises error
```

Comment (gaborh-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: If we want to force the developer to use this feature then we have to do the "obfuscation" underneath, in ledger()/sandbox.
Currently, it's usage is optional and needs to be documented as a best practice, I don't think we have ways to raise error as you write above.

@tamaskalcza-da please weigh in.

Comment (richardkapolnai-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: we have ways, e.g. a `static Set usedPartyNames`.
I'm not sure if we have a fool-proof way to hide this underneath: we'll never know for sure if a party name is "obfuscated" or not. E.g. this is obfuscated:
`party = getContractFromLedgerAndExtractParty(); submitCommand(party);`
but this may not be:
`party = /* handcrafted unique party name, hard to enforce */ ; submitCommand(party);`

Comment (gaborh-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: If we hide it underneath, any function getting a contract should return the "non-obfuscated" party names and all the methods of the API should accept "non-obfuscated" party names as parameters as well.

Comment (gaborh-da) [reply src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java:88]: According to Tomi's explanation how JUnit and similar frameworks work, you cannot easily use `static Set usedPartyNames` to save the list of already used party names.
On the other hand, I still think it is a better idea to ask the user to use our `getUniqueParty` if he wants to parallelize without shooting oneself in the foot. This gives the liberty to one to still use the system without "obfuscation" if there is a need for that.

File: src/test/java/com/digitalasset/testing/PingPongIT.java (hunk 3 above)
Comment (gaborh-da) [reply src/test/java/com/digitalasset/testing/PingPongIT.java]: Done.

**Comment by gaborh-da** [comment]
**Rework reset**
16bb0af3c766df891f031f7abb632c87b243326b
real	8m24,156s
user	12m11,891s
sys	0m22,113s


**origin/master**
c6eafc82ab65968ce3eed4317013c27dd4794bd6
real	15m10,808s
user	18m7,791s
sys	0m34,814s
